server-spi/src/main/java/org/keycloak/models/IDPProvider.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2024 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.models;
+
+import java.util.Map;
+import java.util.stream.Stream;
+
+import org.keycloak.provider.Provider;
+
+/**
+ * The {@code IDPProvider} is concerned with the storage/retrieval of the configured identity providers in Keycloak. In
+ * other words, it is a provider of identity providers (IDPs) and, as such, handles the CRUD operations for IDPs.
+ *
+ * It is not to be confused with the {@code IdentityProvider} found in server-spi-private as that provider is meant to be
+ * implemented by actual identity providers that handle the logic of authenticating users with third party brokers, such
+ * as Microsoft, Google, Github, LinkedIn, etc.
+ *
+ * @author <a href="mailto:sguilhen@redhat.com">Stefan Guilhen</a>
+ */
+public interface IDPProvider extends Provider {
+
+    /**
+     * Creates a new identity provider from the specified model.
+     *
+     * @param model a {@link IdentityProviderModel} containing the identity provider's data.
+     * @return the model of the created identity provider.
+     */
+    IdentityProviderModel create(IdentityProviderModel model);
+
+    /**
+     * Updates the identity provider using the specified model.
+     *
+     * @param model a {@link IdentityProviderModel} containing the identity provider's data.
+     */
+    void updateIdentityProvider(IdentityProviderModel model);
```

Comment:
In other providers there is assumption that the model itself is "live", so that all modifications to it are propagated to the storage layer. Is the IDP different and we plan to update it explicitly via this method?

Reply:
Yeah, we have a different issue with IDPs - the model is extended by several providers, and each of the factories manipulate the model, mixing config that is persisted with config that is runtime-only.

I've discussed with @pedroigor last week and we decided it is best not to touch this for now. It would require a great refactoring to make the IdentityProviderModel capable of handling live updates.

We can already make several performance improvements without changing the model as it is. Caching will end up being simplified too because we can just listen for the update event to run invalidations.